srml/balances/src/lib.rs

```diff
@@ -0,0 +1,1208 @@
+	type Lookup = T::Lookup;
+	type Header = T::Header;
+	type Event = ();
+	type BlockHashCount = T::BlockHashCount;
+	type MaximumBlockWeight = T::MaximumBlockWeight;
+	type MaximumBlockLength = T::MaximumBlockLength;
+	type AvailableBlockRatio = T::AvailableBlockRatio;
+	type Version = T::Version;
+}
+impl<T: Subtrait<I>, I: Instance> timestamp::Trait for ElevatedTrait<T, I> {
+	type Moment = T::Moment;
+	type OnTimestampSet = ();
+	type MinimumPeriod = T::MinimumPeriod;
+}
+impl<T: Subtrait<I>, I: Instance> Trait<I> for ElevatedTrait<T, I> {
+	type Balance = T::Balance;
+	type OnFreeBalanceZero = T::OnFreeBalanceZero;
+	type OnNewAccount = T::OnNewAccount;
+	type TransferPayment = ();
+	type DustRemoval = ();
+	type Event = ();
+	type ExistentialDeposit = T::ExistentialDeposit;
+	type TransferFee = T::TransferFee;
+	type CreationFee = T::CreationFee;
+}
+
+impl<T: Trait<I>, I: Instance> Currency<T::AccountId> for Module<T, I>
+where
+	T::Balance: MaybeSerializeDeserialize + Debug,
+{
+	type Balance = T::Balance;
+	type PositiveImbalance = PositiveImbalance<T, I>;
+	type NegativeImbalance = NegativeImbalance<T, I>;
+
+	fn total_balance(who: &T::AccountId) -> Self::Balance {
+		Self::free_balance(who) + Self::reserved_balance(who)
+	}
+
+	fn can_slash(who: &T::AccountId, value: Self::Balance) -> bool {
+		Self::free_balance(who) >= value
+	}
+
+	fn total_issuance() -> Self::Balance {
+		<TotalIssuance<T, I>>::get()
+	}
+
+	fn minimum_balance() -> Self::Balance {
+		T::ExistentialDeposit::get()
+	}
+
+	fn burn(mut amount: Self::Balance) -> Self::PositiveImbalance {
+		<TotalIssuance<T, I>>::mutate(|issued| {
+			*issued = issued.checked_sub(&amount).unwrap_or_else(|| {
+				amount = *issued;
+				Zero::zero()
+			});
+		});
+		PositiveImbalance::new(amount)
+	}
+
+	fn issue(mut amount: Self::Balance) -> Self::NegativeImbalance {
+		<TotalIssuance<T, I>>::mutate(|issued| {
+			*issued = issued.checked_add(&amount).unwrap_or_else(|| {
+				amount = Self::Balance::max_value() - *issued;
+				Self::Balance::max_value()
+			})
+		});
+		NegativeImbalance::new(amount)
+	}
+
+	fn free_balance(who: &T::AccountId) -> Self::Balance {
+		<FreeBalance<T, I>>::get(who)
+	}
+
+	// # <weight>
+	// Despite iterating over a list of locks, they are limited by the number of
+	// lock IDs, which means the number of runtime modules that intend to use and create locks.
+	// # </weight>
+	fn ensure_can_withdraw(
+		who: &T::AccountId,
+		_amount: T::Balance,
+		reasons: WithdrawReasons,
+		new_balance: T::Balance,
+	) -> Result {
+		if reasons.intersects(WithdrawReason::Reserve | WithdrawReason::Transfer)
+			&& Self::vesting_balance(who) > new_balance
+		{
+			return Err("vesting balance too high to send value");
+		}
+		let locks = Self::locks(who);
+		if locks.is_empty() {
+			return Ok(());
+		}
+
+		let now = <timestamp::Module<T>>::now();
+		if locks
+			.into_iter()
+			.all(|l| now >= l.until || new_balance >= l.amount || !l.reasons.intersects(reasons))
```

Comment:
This condition check here deserve careful checking when multiple locks exists at the same time, e.g. Stash balances and unbonding balances.

It seems that ``` new_balance >= l.amount ``` only checks the largest locked one, so additional locking will not take effect, maybe.